overlord/servicestate/quota_control.go

```diff
@@ -0,0 +1,260 @@
+	"github.com/snapcore/snapd/progress"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/snap/quota"
+	"github.com/snapcore/snapd/snapdenv"
+	"github.com/snapcore/snapd/strutil"
+	"github.com/snapcore/snapd/wrappers"
+)
+
+func ensureSnapServicesForGroup(st *state.State, grp *quota.Group, allGrps map[string]*quota.Group) error {
+	// build the map of snap infos to options to provide to EnsureSnapServices
+	snapSvcMap := map[*snap.Info]*wrappers.SnapServiceOptions{}
+	for _, sn := range grp.Snaps {
+		info, err := snapstate.CurrentInfo(st, sn)
+		if err != nil {
+			return err
+		}
+
+		opts, err := SnapServiceOptions(st, sn, allGrps)
+		if err != nil {
+			return err
+		}
+
+		snapSvcMap[info] = opts
+	}
+
+	// TODO: the following lines should maybe be EnsureOptionsForDevice() or
+	// something since it is duplicated a few places
+	ensureOpts := &wrappers.EnsureSnapServicesOptions{
+		Preseeding: snapdenv.Preseeding(),
+	}
+
+	// set RequireMountedSnapdSnap if we are on UC18+ only
+	deviceCtx, err := snapstate.DeviceCtx(st, nil, nil)
+	if err != nil {
+		return err
+	}
+
+	if !deviceCtx.Classic() && deviceCtx.Model().Base() != "" {
+		ensureOpts.RequireMountedSnapdSnap = true
+	}
+
+	// TODO: do we need to restart modified services ?
+	return wrappers.EnsureSnapServices(snapSvcMap, ensureOpts, nil, progress.Null)
+}
+
+func validateSnapForAddingToGroup(st *state.State, name string, allGrps map[string]*quota.Group) error {
+	// validate that the snap exists
+	_, err := snapstate.CurrentInfo(st, name)
+	if err != nil {
+		return err
+	}
+
+	// check that the snap is not already in a group
+	for _, grp := range allGrps {
+		if strutil.ListContains(grp.Snaps, name) {
+			return fmt.Errorf("snap already in quota group %q", grp.Name)
+		}
+	}
+	return nil
+}
+
+// CreateQuota attempts to create the specified quota group with the specified
+// snaps in it.
+// TODO: should this use something like QuotaGroupUpdate with fewer fields?
+func CreateQuota(st *state.State, name string, parentName string, snaps []string, memoryLimit quantity.Size) error {
+	st.Lock()
+	defer st.Unlock()
+
+	// ensure that the quota group does not exist yet
+	allGrps, err := AllQuotas(st)
+	if err != nil {
+		return err
+	}
+
+	if _, ok := allGrps[name]; ok {
+		return fmt.Errorf("group %q already exists", name)
+	}
+
+	// make sure the specified snaps exist and aren't currently in another group
+	for _, sn := range snaps {
+		if err := validateSnapForAddingToGroup(st, sn, allGrps); err != nil {
+			return fmt.Errorf("cannot use snap %q in group %q: %v", sn, name, err)
+		}
+	}
+
+	// make sure that the parent group exists if we are creating a sub-group
+	var grp, parentGrp *quota.Group
+	if parentName != "" {
+		var ok bool
+		parentGrp, ok = allGrps[parentName]
+		if !ok {
+			return fmt.Errorf("cannot create group under non-existent parent group %q", parentName)
+		}
+
+		grp, err = parentGrp.NewSubGroup(name, memoryLimit)
```

Comment:
maybe a memory limit error case test for this path would be good, can be in a follow-up

Reply:
done in https://github.com/snapcore/snapd/pull/10216